js/serial_manager.js

```diff
@@ -0,0 +1,590 @@
+'use strict';
+var sm;
+var t0;
+var t1;
+$(document).ready(function() {
+	sm = new serial_manager();
+});
+
+var serial_manager = function() {
+	this.port_picker = $('div#port-picker a.connect');
+	this.ac_box = $('input.auto_connect');
+	this.baud_sel = $('div#port-picker #baud');
+	this.startup_message_buffer = "";
+	this.time_of_disconnect = false;
+	this.port_opened_time = false;
+	this.port_open = false;
+	this.init();
+};
+
+serial_manager.prototype.read_serial = function(info) {
+    if (GUI.operating_mode == 3) {
+		// spectrum analyzer
+        SA.read(info);
+	} else {
+		// configurator
+        PSP.read(info);
+    } 
+};
+serial_manager.prototype.send = function(data, callback) {
+    var bufferOut = new ArrayBuffer(data.length),
+        bufferView = new Uint8Array(bufferOut);
+
+    if (typeof data == 'object') {
+        for (var i = 0; i < data.length; i++) {
+            bufferView[i] = data[i];
+        }
+    } else if (typeof data == 'string') {
+        for (var i = 0; i < data.length; i++) {
+            bufferView[i] = data[i].charCodeAt(0);
+        }
+    }
+
+    serial.send(bufferOut, function(writeInfo) {
+        if (writeInfo.bytesSent == bufferOut.byteLength) {
+            if (callback) {
+                callback();
+            }
+        }
+    });
+};
+serial_manager.prototype.fail = function (errID) {
+	var err={
+		connect:{msg:'Failed to open serial port', log:'error_failed_to_open_port'},
+		disconnect:{msg:'Failed to close serial port', log:'error_failed_to_close_port'},
+		no_prog_port:{msg:'no programming port found', log:'error_atmega32u4_programmer_port_not_found'},
+		no_reg_port:{msg:'no regular port found', log:'error_atmega32u4_regular_port_not_found'},
+		no_fw_psp:{msg:'Command: PSP.PSP_REQ_FW_VERSION timed out, connecting failed', log:'error_no_psp_received'},
+		fw_not_supported:{msg:'firmware not supported', log:'firmware_not_supported'}
+	};
+
+	GUI.connecting_to = false;
+	this.port_open = false;
+	this.toggle_port_ctrls(false);
+	
+	var e = err[errID];
+	if(e.msg){
+		console.log(e.msg);
+	}
+	if(e.log){
+		GUI.log(chrome.i18n.getMessage(e.log));
+	}
+};
+serial_manager.prototype.init = function() {
+	var that = this;
```

Comment:
This is a bit odd. `self` is the more commonly used variable name for self references in closures.

Reply:
`that` and `self` are the two most common forms I've always seen for storing the reference to the current scope. I usually use that (just a personal preference), but I can change it to self though, not a big deal... will fix.

Reply:
Doing it consistently within a project has value. Thanks.